### PR TITLE
Scheduler - Appointments overflow groups in certain cases (T1223907)

### PR DESCRIPTION
**Pull request by pomahtri**
Blocked by https://github.com/DevExpress/DevExtreme/pull/27763

### CODE REVIEW

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  const scheduler = new Scheduler('#container');
+  const { takeScreenshot, compareResults } = createScreenshotsComparer(t);
+
+  await t
+    .expect(await takeScreenshot('group-overflow', scheduler.element))
+    .ok()
+    .expect(compareResults.isValid())
+    .ok(compareResults.errorMessages());
+}).before(async () => {
+  await changeTheme('material.blue.light');
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: Let's test it in default theme (generic)
I don't think that we should switch theme here (it's an theme-independent issue)

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+fixture.disablePageReloads`Scheduler: SmoothCellLines`
+  .page(url(__dirname, '../../container.html'));
+
+test('Long appointments should not overflow group view', async (t) => {
+  const scheduler = new Scheduler('#container');
+  const { takeScreenshot, compareResults } = createScreenshotsComparer(t);
+
+  await t
+    .expect(await takeScreenshot('group-overflow', scheduler.element))
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: It's better to write the file extension
```suggestion
    .expect(await takeScreenshot('group-overflow.png', scheduler.element))
```

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  await changeTheme('material.blue.light');
+
+  return createWidget('dxScheduler', {
+    timeZone: 'America/Los_Angeles',
+    dataSource: [
+      {
+        text: '1',
+        priorityId: 1,
+        startDate: new Date('2021-04-19T16:30:00.000Z'),
+        endDate: new Date('2021-04-25T18:30:00.000Z'),
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: Minor comment

It isn't necessary to wrap the Date string into ```new Date```
```suggestion
        startDate: '2021-04-19T16:30:00',
        endDate: '2021-04-25T18:30:00',
```

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  await t
+    .expect(await takeScreenshot('group-overflow', scheduler.element))
+    .ok()
+    .expect(compareResults.isValid())
+    .ok(compareResults.errorMessages());
+}).before(async () => {
+  await changeTheme('material.blue.light');
+
+  return createWidget('dxScheduler', {
+    timeZone: 'America/Los_Angeles',
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: Let's remove the ```timeZone``` option because it's not related to this issue
Note: after removing the ```timeZone``` option appointments' dates should be changed too

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import { createScreenshotsComparer } from 'devextreme-screenshot-comparer';
+import Scheduler from 'devextreme-testcafe-models/scheduler';
+import { changeTheme } from '../../../helpers/changeTheme';
+import { createWidget } from '../../../helpers/createWidget';
+import url from '../../../helpers/getPageUrl';
+
+fixture.disablePageReloads`Scheduler: SmoothCellLines`
+  .page(url(__dirname, '../../container.html'));
+
+test('Long appointments should not overflow group view', async (t) => {
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: Let's add more test cases for the group overflow scenario
I suggest the following matrix:
- grouping horizontal/vertical
- more views (like timeline/week/month)
- interval count (I guess it's better to test it separately, e.g with values 1 & 3)
- allDayPanelMode (at least 'hidden' & 'allDay', but default value ok too)
- startDayHour & endDayHour (two options, without setting these options and with them)

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            color: '#1e90ff',
+          }, {
+            text: 'High Priority',
+            id: 2,
+            color: '#ff9747',
+          },
+          {
+            text: 'Custom',
+            id: 3,
+            color: 'red',
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: Minor comment

red is ```#ff0000``` or shorthand '#f00'

File: packages/devextreme/js/__internal/scheduler/appointments/m_cell_position_calculator.ts [hunk 7]
```diff
@@ -110,6 +110,15 @@ class BaseStrategy {
...
     );
 
+    const groupEdgeIndices = this.viewDataProvider.getGroupEdgeIndices(validGroupIndex);
+    const vMin = this.getCellPosition(
+      {
+        columnIndex: positionByMap.columnIndex,
+        rowIndex: groupEdgeIndices.firstRowIndex,
+      },
+      inAllDayRow && !this.isVerticalGrouping,
+    ).top;
```
Comment (williamvinogradov) [inline packages/devextreme/js/__internal/scheduler/appointments/m_cell_position_calculator.ts]: Let's refactor this code block a little to make it more readable

```suggestion
    const { top: vMin } = this.getCellPosition(
      {
        columnIndex: positionByMap.columnIndex,
        rowIndex: groupEdgeIndices.firstRowIndex,
      },
      inAllDayRow && !this.isVerticalGrouping,
    );
```

File: packages/devextreme/playground/jquery.html [hunk 8]
```diff
@@ -1,5 +1,6 @@
 <!DOCTYPE html>
```
Comment (williamvinogradov) [inline packages/devextreme/playground/jquery.html]: Just reminder
Accidentally added ```jquery.html``` playground changes

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  ['vertical', 'horizontal'].forEach((groupOrientation) => {
+    ['hidden', 'allDay'].forEach((allDayPanelMode) => {
+      [[9, 14, 60], [0, 24, 360]].forEach(([startDayHour, endDayHour, cellDuration]) => {
+        const allParams = `${viewType} ${groupOrientation} ${allDayPanelMode} ${startDayHour} ${endDayHour}`;
+        test(`Long appointments should not overflow group view (${allParams})`, async (t) => {
+          const scheduler = new Scheduler('#container');
+          const { takeScreenshot, compareResults } = createScreenshotsComparer(t);
+
+          await t
+            .expect(await takeScreenshot(`group-overflow (${allParams}).png`, scheduler.element))
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: Minor comment.

It's better to not use the ``` ``` (spaces) because some machines / OS don't handle it properly
But I think it's not a big point in our case :)

File: e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+fixture.disablePageReloads`Scheduler: Grouping overflow`
+  .page(url(__dirname, '../../container.html'));
+
+['week', 'month'].forEach((viewType) => {
+  ['vertical', 'horizontal'].forEach((groupOrientation) => {
+    ['hidden', 'allDay'].forEach((allDayPanelMode) => {
+      [[9, 14, 60], [0, 24, 360]].forEach(([startDayHour, endDayHour, cellDuration]) => {
+        const allParams = `${viewType} ${groupOrientation} ${allDayPanelMode} ${startDayHour} ${endDayHour}`;
+        test(`Long appointments should not overflow group view (${allParams})`, async (t) => {
```
Comment (williamvinogradov) [inline e2e/testcafe-devextreme/tests/scheduler/grouping/overflow.ts]: Super minor comment.

Let's add an empty line here to make test case start more obvious:

**Review by williamvinogradov — DISMISSED** [review]

**Review by williamvinogradov — APPROVED** [review]
